People/sam/cocoascript boxing cleanup

samdeane opened the pull request:
I'm not sure if you'll want to take everything here, as there might be some Sketch specific things (I'm not certain of that, but I'm aware that I haven't had a chance to factor out all the changes nicely). There are certainly some project changes caused by upgrading to XC7/8, for example, and some other older debugging changes.

However, you might want to take an attempt I made to clean up the boxing, and a possible fix for the Mocha crash.

We had been operating with the "fix" that just called `JSValueProtect` on every object, and thus leaked all the JS objects.

I had yet another think about what the problem might be (assuming that it's not a bug in JSCore), and the theory I came up with was that either:
- there might be situations where the JS object gets garbage collected whilst we have a reference to it stored somewhere other than our box (the prime candidate being ffi invocations)
- or that there might be some short-circuiting going on where JSCore doesn't call finalize because it knows that we've made an object which nothing ever references

I suspect that the second situation might be happening with `MOStruct`. We seem to make JS objects here, and thus cause them to be boxed, but we don't actually hold on to the JS refs anywhere that I can see, other than the boxes themselves. The objects seem to be made in passing, as a way of extracting structures out of ffi return values. 

My speculation is that in this situation JSCore might be "smart" enough not to bother actually initialising/finalising the object at all (or it might be a bug that it doesn't get finalised).

My fix for these two possible scenarios is to call `JSValueProtect`:
- on every argument before a function/method invocation
- on every JS object who's corresponding Obj-C object is stored in an `MOStruct`

I can then balance these with calls to `JSValueUnprotect`:
- after the function invocation returns
- when an `MOStruct` is destroyed

I'm not certain that this is a valid fix, but it seems to work in one of the cases that was crashing for me here.

It's a slightly better fix than just calling `JSValueProtect` everywhere, as it does allow some objects to be freed up.

I've also improved the cleanup that COScript/MochaRuntime does when shutting down. In particular, they clear out the box table, and the script removes its reference to the runtime, thus breaking a retain cycle that was happening for us (because, I think, the runtime contained a JS variable that had a reference to the script).

---

**Note**: This is a revision of #39, which was a revision of #34. I'm doing my best to confuse everyone...

samdeane committed:
Changed NSLog to debug to avoid spamming the console in release builds.

samdeane committed:
Changed NSLog to debug to avoid spamming the console in release builds.

samdeane committed:
Merge branch 'warning-fixes'

samdeane committed:
XC6 update

samdeane committed:
Merge branch 'master' of github.com:BohemianCoding/CocoaScript

samdeane committed:
Merge commit 'b8e60ecff73c3907dc80b573fb000d63fa50f956' into HEAD

samdeane committed:
Merge commit 'd265a7d'

samdeane committed:
Merge branch 'master' of github.com:ccgus/CocoaScript

samdeane committed:
Merge branch 'debug-callback'

samdeane committed:
Fixed debug macro.

samdeane committed:
Use a name for the debug method which doesn't clash with the debug macro (doh).

samdeane committed:
Merge branch 'debug-callback'

samdeane committed:
Merge branch 'javascript-debugging-support'

samdeane committed:
Merge branch 'javascript-debugging-support'

samdeane committed:
Merge branch 'javascript-debugging-support'

samdeane committed:
Fixed use of deprecated API

samdeane committed:
Turned #pragma messages into comments

Unfortunately there's no other way to suppress them optionally that I can find, and they were the only warnings in our build.

samdeane committed:
Minor (and slightly ugly) tweak to avoid a warning in release build.

samdeane committed:
Turn on SKIP_INSTALL to avoid a warning when archiving

samdeane committed:
XC6.1 update

samdeane committed:
XC6.1 update - specify the signature for objc_msgSend to prevent a warning

randomsequence committed:
Import: Add support for absolute and ~ paths

Import paths can be in the format:

    relative/to/script.js
    /absolute/path/to/script.js
    ~/relative/to/home/script.js

randomsequence committed:
import: reinstate warning for empty base in non-absolute URLs

randomsequence committed:
Merge pull request #2 from BohemianCoding/feature/3319

BohemianCoding/Sketch#3319 Script import: Add support for absolute and ~ paths

randomsequence committed:
Add support for nested @import statements

Also guards against multiple imports of the same URL, and therefore circular imports

Ref BohemianCoding/Sketch#3319

randomsequence committed:
Merge pull request #3 from BohemianCoding/feature/3319

BohemianCoding/Sketch#3319 Add support for nested @import statements

randomsequence committed:
Fix a bug which ignored first line of @import-ed scripts

randomsequence committed:
Merge pull request #4 from BohemianCoding/feature/3319

Fix a bug which ignored first line of @import-ed scripts

samdeane committed:
Added a test script that exhibits the crash

samdeane committed:
Toned down the debug output for the command line tool whilst debugging the crash.

samdeane committed:
reduced loop as script should crash

samdeane committed:
Slight tweaks to MOBox - tracking down crash

samdeane committed:
Added Andrey's temporary workaround to the crash

samdeane committed:
Perform cleanup before exiting - easier to test if it's working...

samdeane committed:
Can't call unprotect from finalize - it wouldn't make much difference here anyway...

samdeane committed:
Latest changes from Gus:

Merge commit '8ddb9e430f6bc728708ae8090ac70fdf3ba7c24d' into develop

Conflicts:
	src/framework/mocha/MochaRuntime.m

samdeane committed:
Fixed leak of CGColorSpaceRef

samdeane committed:
Keep the analyzer happy with a check for a nil encoding.

samdeane committed:
Fix for a nice leak.

samdeane committed:
Disabled the analyzer for a particularly bonkers piece of code.

samdeane committed:
Another analyzer fix.
Not sure about this one - looks like the wrong size was being used, and the wrong parameter, but I may have misinterpreted things...

samdeane committed:
Disabled the analzyer for some particularly gnarly code.

samdeane committed:
Fixed leak of CGColorSpaceRef

samdeane committed:
Fixed leak of CGColorSpaceRef.
Annotated createImageRefFromBuffer correctly to tell the analyzer that it returns a new thing...

samdeane committed:
Fixed warning.

samdeane committed:
Changed #pragma to comment to stop it generating a message.

samdeane committed:
Merge branch 'analyzer-fixes' into develop

Conflicts:
	src/framework/todparsekit/TDToken.m

samdeane committed:
XC7 updates.

samdeane committed:
Don't use nullable on old versions of Xcode that don't know about it...

samdeane committed:
and another that XC6.2 doesn't know about

samdeane committed:
Slightly cleaner XC6.2 fix.

randomsequence committed:
Replace NSMapTable with NSDictionary & associated objects
Ref BohemianCoding/Sketch#5799

randomsequence committed:
Associate the unboxer with the object, not the box.

randomsequence committed:
Remove objects to box map in favour of directly associated objects

randomsequence committed:
Refactor object to box mapping into MOBox
Ref BohemianCoding/Sketch#5799

randomsequence committed:
Merge pull request #6 from BohemianCoding/feature/5799

Review 5799

randomsequence committed:
Revert to 34fd6d9
Ref BohemianCoding/Sketch#6742

randomsequence committed:
Remove MochaUnboxer, which wouldn't have unboxed anything anyway
Ref BohemianCoding/Sketch#6742

opsGavin committed:
Fixed an analyser issue

samdeane committed:
Added MOBoxManager class to clean up access to boxes a bit.
Added explicit unprotection and removal of boxes in an attempt to prevent reference loops when session is cleaned up.

samdeane committed:
Cleaned up boxing a bit, removed the JSProtect hack.

samdeane committed:
Protect function/selector arguments around calls.
The theory here is that JS references to objects may be in the ffi arguments as raw pointers, so it's a good idea to protect everything. May be spurious...

samdeane committed:
Protect all the JS objects representing members of the struct. These get created but aren't referenced from JS anywhere at the point the struct is made, so I suspect that they might be destroyed again immediately.

samdeane committed:
Make MOStruct unprotect values when it's shut down.

samdeane committed:
Use strong map again instead of dictionary.
We will "leak" any object in the map until the JS reference is garbage collected, or until the end of the script session - which is not ideal, but better than crashing.

samdeane committed:
Simplified slightly.

samdeane committed:
Commented MOBoxManager

samdeane committed:
Cleaned up and remove some obsolete stuff.

samdeane committed:
Clean out runtime when shutting down COScript, to break retain cycle and ensure everything gets disposed of.

samdeane committed:
Strengthened assertion.

samdeane committed:
Brought up to date with Gus' master branch.

# Conflicts:
#	Cocoa Script.xcodeproj/project.pbxproj
#	src/framework/mocha/MochaRuntime.m
#	src/framework/mocha/Objects/MOBox.m

samdeane committed:
A bit more cleanup.

samdeane committed:
Added some more assertions.

samdeane committed:
Only add box to the index if the JS initialiser actually gets called.

samdeane committed:
More debugging.

samdeane committed:
Don't make a box at all until the JS object has been initialized.
Make sure that boxes get dissassociated and clean out the JS object's private pointer, before they are dealloced.

samdeane committed:
improved comments and removed some more debug

samdeane committed:
Cleaner implementation of the box creation - pass parameters in a custom class instead of a dictionary.

samdeane committed:
Fixed out of date comment.

samdeane committed:
Disabled some additional debugging.

samdeane committed:
improved comment

samdeane committed:
Disable assertions in release builds.

samdeane committed:
Downgraded an assertion to logging instead.

ccgus commented:
Finally getting around to this.
Why main thread for MOBoxManager initWithContext? Just to make things simpler?

samdeane commented:
Actually, I don't think the thread matters per-se, I was just trying to check that the box manager wasn't being hit from multiple threads. Probably it should record the thread that it was initialised on, and just assert that it's only used from that thread. Could also make it thread-safe of course, but I was wondering if it was some sort of race condition causing the crash, so the checks were just a little bit of investigation.

samdeane committed:
Analyzer warning fixes for situations where a parameter might be nil.

samdeane committed:
Supply a name for the context when creating a COScript object.

samdeane committed:
made JSTTextView part of Cocoascript so that we can use it in Sketch

samdeane committed:
Removed some deprecated API usage.

samdeane committed:
Pass back the script URL to the error handler if we have it.

samdeane committed:
Merge commit 'eac11a0' into feature/8204-improve-editing-inside

samdeane committed:
Defer re-parsing of text to avoid messing up the selection.

samdeane committed:
Fix for endless re-parsing problem.

samdeane committed:
Updated for XC8.

samdeane committed:
Turned off the localisation warning that XC8 enabled.

samdeane committed:
Ensure that runtime is thrown away when we cleanup - breaks retain cycle which would otherwise occur if the runtime contains a global object that owns a reference to the COScript object (which happens in Sketch).

samdeane committed:
Merge branch 'develop' into feature/9844-memory-leak-in-plugins-

samdeane committed:
Merge branch 'feature/9844-memory-leak-in-plugins-' into people/sam/cocoascript-boxing-cleanup

# Conflicts:
#	Cocoa Script.xcodeproj/project.pbxproj
#	src/framework/COScript.m

samdeane committed:
Factored out the code that calls JSObjectGetPrivate, and that returns the boxed object, so that we can annotate it with extra debug if necessary.

samdeane committed:
added a generation counter, and some more debugging

samdeane committed:
Cleaned up some index behaviour.
We were sometimes missing out some boxes when performing cleanup, because the keys returned to us by (for NSValue* key in _index) were failing to return a value when [_index objectForKey:key] was called!
I presume that this is because the key lookup was using isEqual: and somehow was failing to match somehow.
To avoid this we turn on NSMapTableObjectPointerPersonality for the keys.
For extra paranoia, we iterate the collection's objects, rather than its keys, so that we can be certain we're getting them all.

samdeane committed:
Fixed analyser warning.

samdeane commented:
I've been looking at this yet again, in relation to another problem we had.

Our crude workaround for the crash (not this branch, but in the branch that Sketch is currently using) was just to leak all the JS objects until we cleaned up the context after the script executed.

However at some point we created a retain cycle so we weren't actually cleaning up the context ever - not good. In fixing that I uncovered another problem with cleanup, so I thought I'd give this branch another try whilst debugging it. 

The cleanup problem still happened on this branch, and I tracked it down to an issue with the lookup table that keeps the boxed objects, which was failing to look up the box sometimes because it wasn't using pointer identity for the keys.

samdeane commented:
That pointer identity fix was something that Logan has actually incorporated into Mocha now, but when I rewrote the boxing stuff in this branch I missed it.

I'm still not sure if this branch will actually fix the main problem we're seeing, but I'll let you know.
